// airbnb-private-api/lib/constant/index.js
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.user_agents = [
    {
        user_agent: 'Airbnb/21057117 AppVersion/20.07.1 Android/10.0 Device/Pixel 2 XL Carrier/US T-MOBILE Type/Phone',
        screen_size: 'w=360.0;h=640.0',
    },
];
exports.maint_host = 'https://api.airbnb.com/v2';
exports.api_key = '3092nxybyb0otqw18e8nh5nty';
exports.user_types = ['guest', 'host', 'experience_host', 'guest_and_host'];
exports.reservation_order = ['start_date', 'nights', 'number_of_guests', 'status'];
exports.endpoints = {
    authentications: `${exports.maint_host}/authentications`,
    messaging_syncs: `${exports.maint_host}/messaging_syncs`,
    threads: id => (id ? `${exports.maint_host}/threads/${id}` : `${exports.maint_host}/threads`),
    accounts_me: `${exports.maint_host}/accounts/me`,
    user_by_id: id => `${exports.maint_host}/users/${id}`,
    wishlists: `${exports.maint_host}/wishlists`,
    listings: id => (id ? `${exports.maint_host}/listings/${id}` : `${exports.maint_host}/listings`),
    host_pricing_calculators: `${exports.maint_host}/host_pricing_calculators`,
    calendar_operations: `${exports.maint_host}/calendar_operations`,
    reservations: `${exports.maint_host}/reservations`,
    homes_booking_details: id => `${exports.maint_host}/homes_booking_details/${id}`,
    phone_one_time_passwords: `${exports.maint_host}/phone_one_time_passwords`,
    explore_tabs: `${exports.maint_host}/explore_tabs`,
};
